internals/overlord/servstate/workloads.go

```diff
@@ -0,0 +1,209 @@
+		dec := yaml.NewDecoder(bytes.NewReader(yml))
+		dec.KnownFields(true)
+		if err = dec.Decode(ws); err != nil {
+			return nil, &plan.FormatError{
+				Message: fmt.Sprintf(`cannot parse the "workloads" section: %v`, err),
+			}
+		}
+	}
+	for name, workload := range ws.Entries {
+		if workload != nil {
+			workload.Name = name
+		}
+	}
+	return ws, nil
+}
+
+func (ext *WorkloadsSectionExtension) ValidatePlan(p *plan.Plan) error {
+	ws, ok := p.Sections[WorkloadsField].(*WorkloadsSection)
+	if !ok {
+		return fmt.Errorf("internal error: invalid section type %T", ws)
+	}
+	for name, service := range p.Services {
+		_, ok := ws.Entries[service.Workload]
+		if service.Workload != "" && !ok {
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`plan service %q cannot run in unknown workload %q`, name, service.Workload),
+			}
+		}
+	}
+	return nil
+}
+
+const WorkloadsField = "workloads"
+
+var _ plan.Section = (*WorkloadsSection)(nil)
+
+type WorkloadsSection struct {
+	Entries map[string]*Workload `yaml:",inline"`
+}
+
+func (ws *WorkloadsSection) IsZero() bool {
+	return len(ws.Entries) == 0
+}
+
+func (ws *WorkloadsSection) Validate() error {
+	for name, workload := range ws.Entries {
+		if workload == nil {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("workload %q has a null value", name),
+			}
+		}
+		if err := workload.validate(); err != nil {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("workload %q %v", name, err),
+			}
+		}
+	}
+	return nil
+}
+
+func (ws *WorkloadsSection) combine(other *WorkloadsSection) error {
+	if len(other.Entries) != 0 && ws.Entries == nil {
+		ws.Entries = make(map[string]*Workload)
+	}
+	for name, workload := range other.Entries {
+		switch workload.Override {
+		case plan.MergeOverride:
+			if current, ok := ws.Entries[name]; ok {
+				copied := current.copy()
+				copied.merge(workload)
+				ws.Entries[name] = copied
+				break
+			}
+			fallthrough
+		case plan.ReplaceOverride:
+			ws.Entries[name] = workload.copy()
+		case plan.UnknownOverride:
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`workload %q must define an "override" policy`, name),
+			}
+		default:
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`workload %q has an invalid "override" policy: %q`, name, workload.Override),
+			}
+		}
+	}
+	return nil
+}
+
+type Workload struct {
+	// Basic details
+	Name     string        `yaml:"-"`
+	Override plan.Override `yaml:"override,omitempty"`
+
+	// Options for command execution
+	Environment map[string]string `yaml:"environment,omitempty"`
+	UserID      *int              `yaml:"user-id,omitempty"`
+	User        string            `yaml:"user,omitempty"`
+	GroupID     *int              `yaml:"group-id,omitempty"`
+	Group       string            `yaml:"group,omitempty"`
+}
+
+func (w *Workload) validate() error {
+	if w.Name == "" {
+		return errors.New("cannot have an empty name")
+	}
+	// Value of Override is checked in the (*WorkloadSection).combine() method
```

Comment:
My intuition is that `validate` should do that check regardless.

Reply:
The way is has been implemented in the plan library is that the combine logic checks this, and therefore the extension followed the same flow. Since both layer creation and merging of the layer has to be successful for a new plan to be updated, it does not really matter in practice where its done, the issue will be caught.

Having said that, since ```plan.Override``` is the same type being used everywhere, it would be super simple to get the type to validate itself as part of layer validation instead, and it could make the ```combine``` a little simpler.

Reply:
It would be trivial to add a `func (plan.Override) Validate() error` method, but it would serve no practical purpose because the validation here is done at the same time as we perform the logic for combining the layers

Reply:
Ok, please resolve